Fix Windows-specific test failures in features/workflow package

## Issue Description

Windows CI builds are failing in the `features/workflow` package due to async logging timing issues. Tests pass on Linux and macOS.

## Current Status

- **Affected**: Windows builds only
- **Impact**: Blocking CI on Windows (pre-existing issue)
- **Tests Affected**: 
  - `TestWorkflowConditionalExecution`
  - `TestDebugEngine_SessionManagement`
  - Other tests with error validation

## Root Cause

Tests that validate error handling use `pkgtesting.NewMockLogger(true)` which prints expected error logs. On Windows, timing differences cause these error logs to appear after test completion, and the CI interprets stderr output as test failures.

**Example error logs** (these are EXPECTED from error handling tests):
```
[ERROR] Workflow execution failed: HTTP configuration is required for HTTP steps
[ERROR] Workflow execution failed: barrier timeout after 10ms
[ERROR] Workflow execution failed: try step missing try configuration
```

## Evidence

1. **Windows CI**: Tests fail consistently
   - Run: https://github.com/cfg-is/cfgms/actions/runs/20804871289/job/59757154564
   
2. **Linux/macOS**: Same tests pass
   - Verified locally: Tests complete successfully
   
3. **Pre-existing**: Also fails on develop branch
   - Not a regression from recent changes

## Proposed Solutions

### Option 1: Fix Async Logging (Recommended)
- Ensure all async goroutines complete before test exits
- Add proper cleanup/wait mechanisms in test teardown
- Prevent race between test completion and log output

### Option 2: Suppress Expected Error Logs in Tests
- Use `pkgtesting.NewMockLogger(false)` for tests that validate errors
- Only enable verbose logging for debugging specific issues
- Reduces noise in CI output

### Option 3: Platform-Specific Test Behavior
- Add build tag for problematic tests
- Create Windows-specific versions if needed
- Not ideal but unblocks CI quickly

## Acceptance Criteria

- [ ] All `features/workflow` tests pass on Windows CI
- [ ] Tests continue to pass on Linux and macOS
- [ ] No race conditions between test completion and logging
- [ ] Error handling validation tests still work correctly

## Priority

**Medium** - Pre-existing issue that doesn't affect production code, but blocks Windows CI builds.

## Related

- Originally discovered during PR #308 review
- Affects cross-platform build validation workflow

--- a/features/workflow/debug_test.go
+++ b/features/workflow/debug_test.go
@@ -370,7 +370,8 @@ func TestDebugEngine_SessionManagement(t *testing.T) {
 	require.NoError(t, err)
 
 	// Wait for workflows to initialize (Windows CI timing)
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution1, 2*time.Second)
+	waitForWorkflowCompletion(t, execution2, 2*time.Second)
 
 	// Start debug sessions
 	settings := DebugSettings{MaxHistorySize: 100}

--- a/features/workflow/engine_test.go
+++ b/features/workflow/engine_test.go
@@ -82,7 +82,7 @@ func TestEngine_ExecuteWorkflow_Simple(t *testing.T) {
 	assert.NotEmpty(t, execution.ID)
 
 	// Wait for execution to complete
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Check final status
 	finalExecution, err := engine.GetExecution(execution.ID)
@@ -137,7 +137,7 @@ func TestEngine_ExecuteWorkflow_Parallel(t *testing.T) {
 	require.NoError(t, err)
 
 	// Wait for execution to complete
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	finalExecution, err := engine.GetExecution(execution.ID)
 	require.NoError(t, err)

--- a/features/workflow/error_workflow_test.go
+++ b/features/workflow/error_workflow_test.go
@@ -52,8 +52,8 @@ func TestErrorWorkflowStep(t *testing.T) {
 	require.NoError(t, err)
 	require.NotNil(t, execution)
 
-	// Wait for completion
-	time.Sleep(200 * time.Millisecond)
+	// Wait for completion (fixes Windows CI async race condition - Issue #309)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -96,8 +96,8 @@ func TestErrorWorkflowWithPath(t *testing.T) {
 	require.NoError(t, err)
 	require.NotNil(t, execution)
 
-	// Wait for completion
-	time.Sleep(200 * time.Millisecond)
+	// Wait for completion (fixes Windows CI async race condition - Issue #309)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -152,8 +152,8 @@ func TestErrorWorkflowAsync(t *testing.T) {
 	require.NoError(t, err)
 	require.NotNil(t, execution)
 
-	// Wait for completion
-	time.Sleep(200 * time.Millisecond)
+	// Wait for completion (fixes Windows CI async race condition - Issue #309)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -188,7 +188,7 @@ func TestErrorWorkflowMissingConfiguration(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution failed due to missing configuration
 	assert.Equal(t, StatusFailed, execution.GetStatus())
@@ -223,7 +223,7 @@ func TestErrorWorkflowMissingWorkflowSpec(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution failed due to missing workflow specification
 	assert.Equal(t, StatusFailed, execution.GetStatus())
@@ -269,7 +269,7 @@ func TestErrorWorkflowWithRecoveryActions(t *testing.T) {
 			require.NotNil(t, execution)
 
 			// Wait for completion
-			time.Sleep(200 * time.Millisecond)
+			waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 			// Verify execution completed successfully
 			assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -326,8 +326,8 @@ func TestErrorWorkflowParameterAndOutputMappings(t *testing.T) {
 	require.NoError(t, err)
 	require.NotNil(t, execution)
 
-	// Wait for completion
-	time.Sleep(200 * time.Millisecond)
+	// Wait for completion (fixes Windows CI async race condition - Issue #309)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -368,8 +368,8 @@ func TestErrorWorkflowTimeout(t *testing.T) {
 	require.NoError(t, err)
 	require.NotNil(t, execution)
 
-	// Wait for completion
-	time.Sleep(200 * time.Millisecond)
+	// Wait for completion (fixes Windows CI async race condition - Issue #309)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// The workflow should complete (the timeout handling is internal to the error workflow)
 	// The actual timeout behavior depends on the implementation details

--- a/features/workflow/fanout_fanin_test.go
+++ b/features/workflow/fanout_fanin_test.go
@@ -57,7 +57,7 @@ func TestFanOutStep(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -106,7 +106,7 @@ func TestFanInStep(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -210,7 +210,7 @@ func TestFanInStrategies(t *testing.T) {
 			require.NotNil(t, execution)
 
 			// Wait for completion
-			time.Sleep(100 * time.Millisecond)
+			waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 			// Verify execution completed successfully
 			assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -259,7 +259,7 @@ func TestFanOutEmptyData(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully (empty fan-out is valid)
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -297,7 +297,7 @@ func TestFanInEmptyData(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -332,7 +332,7 @@ func TestFanOutMissingConfiguration(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution failed due to missing configuration
 	assert.Equal(t, StatusFailed, execution.GetStatus())
@@ -362,7 +362,7 @@ func TestFanInMissingConfiguration(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution failed due to missing configuration
 	assert.Equal(t, StatusFailed, execution.GetStatus())
@@ -415,7 +415,7 @@ func TestFanOutFanInCombined(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(200 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())

--- a/features/workflow/http_test.go
+++ b/features/workflow/http_test.go
@@ -135,7 +135,7 @@ func TestEngine_ExecuteHTTPStep(t *testing.T) {
 	require.NoError(t, err)
 
 	// Wait for execution to complete
-	time.Sleep(200 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	finalExecution, err := engine.GetExecution(execution.ID)
 	require.NoError(t, err)
@@ -202,7 +202,7 @@ func TestEngine_ExecuteAPIStep(t *testing.T) {
 	require.NoError(t, err)
 
 	// Wait for execution to complete
-	time.Sleep(200 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	finalExecution, err := engine.GetExecution(execution.ID)
 	require.NoError(t, err)
@@ -265,7 +265,7 @@ func TestEngine_ExecuteWebhookStep(t *testing.T) {
 	require.NoError(t, err)
 
 	// Wait for execution to complete
-	time.Sleep(200 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	finalExecution, err := engine.GetExecution(execution.ID)
 	require.NoError(t, err)
@@ -306,7 +306,7 @@ func TestEngine_ExecuteDelayStep(t *testing.T) {
 	require.NoError(t, err)
 
 	// Wait for execution to complete
-	time.Sleep(300 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	finalExecution, err := engine.GetExecution(execution.ID)
 	require.NoError(t, err)
@@ -415,7 +415,7 @@ func TestEngine_ComplexAPIWorkflow(t *testing.T) {
 	require.NoError(t, err)
 
 	// Wait for execution to complete
-	time.Sleep(500 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	finalExecution, err := engine.GetExecution(execution.ID)
 	require.NoError(t, err)

--- a/features/workflow/loop_break_test.go
+++ b/features/workflow/loop_break_test.go
@@ -60,7 +60,7 @@ func TestForLoopBreak(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -118,7 +118,7 @@ func TestForLoopContinue(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -181,7 +181,7 @@ func TestWhileLoopBreak(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -238,7 +238,7 @@ func TestForeachLoopBreak(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -276,7 +276,7 @@ func TestBreakContinueOutsideLoop(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// The workflow should complete, but the break step should be treated as an error
 	// Since we're outside a loop, the LoopControlError should propagate as a normal error

--- a/features/workflow/ml_logging_test.go
+++ b/features/workflow/ml_logging_test.go
@@ -513,7 +513,7 @@ func TestMLLogger_BufferManagement(t *testing.T) {
 	mlLogger.LogPerformanceSnapshot(execution, "step3") // Should trigger flush
 
 	// Wait a bit for async flush
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify entries were flushed to provider
 	entries := mockProvider.GetLoggedEntries()

--- a/features/workflow/nested_workflow_test.go
+++ b/features/workflow/nested_workflow_test.go
@@ -48,7 +48,7 @@ func TestNestedWorkflowByName(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -94,7 +94,7 @@ func TestNestedWorkflowByPath(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -142,7 +142,7 @@ func TestNestedWorkflowParameterMapping(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -176,7 +176,7 @@ func TestNestedWorkflowTimeout(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// This might complete successfully if the nested workflow executes very quickly,
 	// or fail due to timeout. Both are acceptable outcomes for this test.
@@ -220,7 +220,7 @@ func TestNestedWorkflowAsync(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -250,7 +250,7 @@ func TestNestedWorkflowMissingConfiguration(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution failed due to missing configuration
 	assert.Equal(t, StatusFailed, execution.GetStatus())
@@ -285,7 +285,7 @@ func TestNestedWorkflowMissingWorkflowSpec(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution failed due to missing workflow specification
 	assert.Equal(t, StatusFailed, execution.GetStatus())
@@ -317,7 +317,7 @@ func TestNestedWorkflowNotFound(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution failed due to workflow not found
 	assert.Equal(t, StatusFailed, execution.GetStatus())
@@ -365,7 +365,7 @@ func TestNestedWorkflowComplexParameterMapping(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())

--- a/features/workflow/providers_test.go
+++ b/features/workflow/providers_test.go
@@ -244,7 +244,7 @@ func TestEngine_ExecuteAPIStep_WithProviderRegistry(t *testing.T) {
 	require.NoError(t, err)
 
 	// Wait for execution to complete
-	time.Sleep(200 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	finalExecution, err := engine.GetExecution(execution.ID)
 	require.NoError(t, err)

--- a/features/workflow/switch_test.go
+++ b/features/workflow/switch_test.go
@@ -162,7 +162,7 @@ func TestSwitchStep_BasicFunctionality(t *testing.T) {
 			require.NotNil(t, execution)
 
 			// Wait for completion
-			time.Sleep(100 * time.Millisecond)
+			waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 			// Verify execution completed successfully
 			assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -243,7 +243,7 @@ func TestSwitchStep_DefaultCase(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -297,7 +297,7 @@ func TestSwitchStep_NoMatchNoDefault(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully (no match is valid)
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -376,7 +376,7 @@ func TestSwitchStep_ExpressionBasedSwitch(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -426,7 +426,7 @@ func TestSwitchStep_VariableResolutionError(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution failed due to missing variable
 	assert.Equal(t, StatusFailed, execution.GetStatus())
@@ -506,7 +506,7 @@ func TestSwitchStep_NestedSwitchStatements(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())

--- a/features/workflow/sync_test.go
+++ b/features/workflow/sync_test.go
@@ -103,7 +103,7 @@ func TestBarrierStep(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(200 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -159,7 +159,7 @@ func TestSemaphoreStep(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -241,7 +241,7 @@ func TestLockStep(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -345,7 +345,7 @@ func TestWaitGroupStep(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(200 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution completed successfully
 	assert.Equal(t, StatusCompleted, execution.GetStatus())
@@ -408,7 +408,7 @@ func TestSyncMissingConfiguration(t *testing.T) {
 			require.NotNil(t, execution)
 
 			// Wait for completion
-			time.Sleep(100 * time.Millisecond)
+			waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 			// Verify execution failed due to missing configuration
 			assert.Equal(t, StatusFailed, execution.GetStatus())
@@ -444,7 +444,7 @@ func TestSyncTimeout(t *testing.T) {
 	require.NotNil(t, execution)
 
 	// Wait for completion
-	time.Sleep(100 * time.Millisecond)
+	waitForWorkflowCompletion(t, execution, 2*time.Second)
 
 	// Verify execution failed due to timeout
 	assert.Equal(t, StatusFailed, execution.GetStatus())
@@ -467,7 +467,7 @@ func TestSyncCleanup(t *testing.T) {
 
 	// Complete one barrier
 	go func() {
-		time.Sleep(10 * time.Millisecond)
+		time.Sleep(50 * time.Millisecond)
 		_ = barrier1.Wait(context.Background(), 100*time.Millisecond)
 	}()
 	_ = barrier1.Wait(context.Background(), 100*time.Millisecond)

--- a/features/workflow/test_helpers.go
+++ b/features/workflow/test_helpers.go
@@ -1,0 +1,48 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2026 Jordan Ritz
+package workflow
+
+import (
+	"testing"
+	"time"
+)
+
+// waitForWorkflowCompletion polls the workflow execution status until it reaches
+// a terminal state (completed, failed, or cancelled) or times out.
+//
+// This is the correct pattern for testing async workflow execution, replacing
+// fixed time.Sleep() calls that cause race conditions on slower systems (Windows CI).
+//
+// Note: This function polls the execution object's status directly. The workflow
+// engine updates the status field atomically, so we can poll without fetching
+// from the engine each time.
+//
+// Usage:
+//
+//	execution, err := engine.ExecuteWorkflow(ctx, workflow, nil)
+//	require.NoError(t, err)
+//	waitForWorkflowCompletion(t, execution, 2*time.Second)
+//	assert.Equal(t, StatusCompleted, execution.GetStatus())
+func waitForWorkflowCompletion(t *testing.T, execution *WorkflowExecution, timeout time.Duration) {
+	t.Helper()
+
+	deadline := time.Now().Add(timeout)
+	ticker := time.NewTicker(10 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			// Use GetStatus() which properly locks and reads the status
+			status := execution.GetStatus()
+
+			if status != StatusPending && status != StatusRunning {
+				return
+			}
+			if time.Now().After(deadline) {
+				t.Fatalf("timeout waiting for workflow completion after %v, status: %s, execution_id: %s",
+					timeout, status, execution.ID)
+			}
+		}
+	}
+}

--- a/features/workflow/try_catch_test.go
+++ b/features/workflow/try_catch_test.go
@@ -13,29 +13,6 @@ import (
 	pkgtesting "github.com/cfgis/cfgms/pkg/testing"
 )
 
-// waitForWorkflowCompletion polls the workflow execution until it completes or times out
-func waitForWorkflowCompletion(t *testing.T, execution *WorkflowExecution, timeout time.Duration) {
-	t.Helper()
-	deadline := time.Now().Add(timeout)
-	ticker := time.NewTicker(10 * time.Millisecond)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case <-ticker.C:
-			status := execution.GetStatus()
-			if status != StatusPending && status != StatusRunning {
-				return
-			}
-		case <-time.After(time.Until(deadline)):
-			if time.Now().After(deadline) {
-				status := execution.GetStatus()
-				t.Fatalf("Workflow did not complete within %v, final status: %v", timeout, status)
-			}
-		}
-	}
-}
-
 func TestTrySuccess(t *testing.T) {
 	// Test try block that succeeds - finally should run, catch should not
 	workflow := Workflow{

